### PR TITLE
storage: aggregate iterator stats

### LINKED ISSUE #95790: storage: consider aggregating iterator stats
Pebble Iterators expose iterator statistics. These statistics are included in SQL traces, and can be helpful in determining where reads are spending time. We can consider aggregating these iterator statistics across all of a node's iterators, to surface time series metrics. These metrics might be helpful in diagnosing problems where slow/expensive reads are substantial enough to affect node health.

Jira issue: CRDB-23742

### PR DESCRIPTION
**Pull request by jbowens**
Aggregate the iterator stats across all of an engine's iterators. Expose seven
new timeseries metrics for visibility into the behavior of storage engine iterators:

  - storage.iterator.block-load.bytes
  - storage.iterator.block-load.cached-bytes
  - storage.iterator.block-load.read-duration
  - storage.iterator.external.seeks
  - storage.iterator.external.steps
  - storage.iterator.internal.seeks
  - storage.iterator.internal.steps

Close https://github.com/cockroachdb/cockroach/issues/95790.
Epic: None
Release note (ops change): Introduces seven new timeseries metrics for better
visibility into the behavior of storage engine iterators and their internals.

### CODE REVIEW

File: pkg/kv/kvserver/metrics.go [hunk 1]
```diff
@@ -2421,6 +2470,13 @@ func newStoreMetrics(histogramWindow time.Duration) *StoreMetrics {
 		RdbLevelScore:               rdbLevelScore,
 		RdbWriteStalls:              metric.NewGauge(metaRdbWriteStalls),
 		RdbWriteStallNanos:          metric.NewGauge(metaRdbWriteStallNanos),
+		IterBlockBytes:              metric.NewGauge(metaBlockBytes),
```
Comment (tbg) [inline pkg/kv/kvserver/metrics.go:2473]: I was expecting them all to be counters. They are all gauges. What's going on?

File: pkg/storage/engine.go [hunk 2]
```diff
@@ -1037,6 +1031,56 @@ type Metrics struct {
...
+	//
+	// We do not split this metric across these two reasons, but they can be
+	// distinguished in the pebble logs.
+	WriteStallCount    int64
+	WriteStallDuration time.Duration
+}
+
+// AggregatedIteratorStats holds cumulative stats, collected and summed over all
+// of an engine's iterators.
+type AggregatedIteratorStats struct {
```
Comment (tbg) [inline pkg/storage/engine.go:1047]: It would be helpful to reference this type in the help text for each of the metrics as a pointer for more detailed semantics of what's tracked within each.

File: pkg/storage/engine.go [hunk 3]
```diff
@@ -1037,6 +1031,56 @@ type Metrics struct {
...
+	BlockReadDuration time.Duration
+	// ExternalSeeks is the total count of seeks in forward and backward
+	// directions performed on pebble.Iterators.
+	ExternalSeeks int
+	// ExternalSteps is the total count of relative positioning operations (eg,
+	// Nexts, Prevs, NextPrefix, NextWithLimit, etc) in forward and backward
+	// directions performed on pebble.Iterators.
+	ExternalSteps int
+	// InternalSeeks is the total count of steps in forward and backward
+	// directions performed on Pebble's internal iterator. If this is high
```
Comment (tbg) [inline pkg/storage/engine.go:1074]: "If this is high..." could you copy this to the metric help text?

File: pkg/storage/engine.go [hunk 4]
```diff
@@ -1037,6 +1031,56 @@ type Metrics struct {
...
+	// relative to ExternalSeeks, it's a good indication that there's an
+	// accumulation of garbage within the LSM (NOT MVCC garbage).
+	InternalSeeks int
+	// InternalSteps is the total count of relative positioning operations (eg,
+	// Nexts, Prevs, NextPrefix, etc) in forward and backward directions
+	// performed on pebble's internal iterator. If this is high relative to
+	// ExternalSteps, it's a good indication that there's an accumulation of
+	// garbage within the LSM (NOT MVCC garbage).
```
Comment (tbg) [inline pkg/storage/engine.go:1082]: ditto about "If this is high..."

File: pkg/storage/pebble_iterator.go [hunk 5]
```diff
@@ -46,6 +46,11 @@ type pebbleIterator struct {
 	// Buffer used to store MVCCRangeKeyVersions returned by RangeKeys(). Lazily
 	// initialized the first time an iterator's RangeKeys() method is called.
 	mvccRangeKeyVersions []MVCCRangeKeyVersion
+	// reportStats is a function that should be invoked when the iterator is
+	// closed or its stats reset. It's intended to be used with
+	// Pebble.aggregateIterStats to sum iterator stats across all the iterators
+	// during the lifetime of the Engine.
```
Comment (tbg) [inline pkg/storage/pebble_iterator.go:52]: and it can't be nil, right?

**Review by tbg — COMMENTED** [review]
Looks good! Had some comments, the most significant ones are

- lock contention, which we might be catching a lot of here, and
- whether these metrics should all be counters.

An additional expert review on the smaller code changes when passing the stats around wouldn't hurt, though it looked straightforward enough.

**Review by jbowens — COMMENTED** [review]
> lock contention, which we might be catching a lot of here, and
> whether these metrics should all be counters.

Yeah, these two are related. I believe using a mutex _without_ a standard `metric.Counter` (which internally uses atomic intrinsics) will be faster and have less contention. The critical section here is so tiny that the cost of synchronization itself should dominate. 

A single atomic increment intrinsic is faster than a mutex lock, add, and mutex unlock, but my understanding is that atomic intrinsics pay the cost of synchronization (eg, write barriers, cache line invalidation) per increment. Since we have seven counters to increment, the mutex overhead gets amortized over the seven increments.

eg, running [this benchmark](https://gist.github.com/jbowens/1c75cdcf8baf19001c166ee4092e0829) on my gceworker:
```
goos: linux
goarch: amd64
pkg: github.com/mutex
cpu: Intel(R) Xeon(R) CPU @ 2.30GHz
BenchmarkInc
BenchmarkInc/atomic
BenchmarkInc/atomic/parallelism=1
BenchmarkInc/atomic/parallelism=1-24         	25120298	        47.57 ns/op
BenchmarkInc/atomic/parallelism=2
BenchmarkInc/atomic/parallelism=2-24         	 4960356	       255.0 ns/op
BenchmarkInc/atomic/parallelism=4
BenchmarkInc/atomic/parallelism=4-24         	 2159403	       554.2 ns/op
BenchmarkInc/atomic/parallelism=8
BenchmarkInc/atomic/parallelism=8-24         	 1000000	      1140 ns/op
BenchmarkInc/atomic/parallelism=16
BenchmarkInc/atomic/parallelism=16-24        	  476132	      2577 ns/op
BenchmarkInc/atomic/parallelism=32
BenchmarkInc/atomic/parallelism=32-24        	  229970	      5171 ns/op
BenchmarkInc/atomic/parallelism=64
BenchmarkInc/atomic/parallelism=64-24        	  117676	     11069 ns/op
BenchmarkInc/mutex
BenchmarkInc/mutex/parallelism=1
BenchmarkInc/mutex/parallelism=1-24          	49485908	        24.13 ns/op
BenchmarkInc/mutex/parallelism=2
BenchmarkInc/mutex/parallelism=2-24          	19753113	        62.67 ns/op
BenchmarkInc/mutex/parallelism=4
BenchmarkInc/mutex/parallelism=4-24          	 5321437	       229.2 ns/op
BenchmarkInc/mutex/parallelism=8
BenchmarkInc/mutex/parallelism=8-24          	 1689697	       701.5 ns/op
BenchmarkInc/mutex/parallelism=16
BenchmarkInc/mutex/parallelism=16-24         	  693480	      1733 ns/op
BenchmarkInc/mutex/parallelism=32
BenchmarkInc/mutex/parallelism=32-24         	  313561	      4045 ns/op
BenchmarkInc/mutex/parallelism=64
BenchmarkInc/mutex/parallelism=64-24         	  144046	      8279 ns/op
```

Also, today _all_ metrics returned from `Engine.GetMetrics()` are gauges because all the metrics reported by Pebble are cumulative, not deltas, even if internally they're monotonically increasing counters.

I goofed and wasn't propagating the metrics in `Engine.GetMetrics()`, so I made that change too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-:-NRe8bjX-COOJLsnsXln:bexvjh9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick and @tbg)

___
*[`pkg/storage/engine.go` line 1047 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJg8cg9w4GYayEs-r2-1047:-NRe7x791RcIGoB_DSIz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d725ca9a6e8b43dae337103a2406e48ff9ea347b/pkg/storage/engine.go#L1047)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It would be helpful to reference this type in the help text for each of the metrics as a pointer for more detailed semantics of what's tracked within each.
</blockquote></details>

Done.
___
*[`pkg/storage/engine.go` line 1074 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJg8cg9w4GYayEs-r2-1074:-NRe8Iap2ZkMtVYFlCAr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d725ca9a6e8b43dae337103a2406e48ff9ea347b/pkg/storage/engine.go#L1074)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

"If this is high..." could you copy this to the metric help text?
</blockquote></details>

Done.
___
*[`pkg/storage/engine.go` line 1082 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJg8cg9w4GYayEs-r2-1082:-NRe8J6bAAHHswztC1Fk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d725ca9a6e8b43dae337103a2406e48ff9ea347b/pkg/storage/engine.go#L1082)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

ditto about "If this is high..."
</blockquote></details>

Done.
___
*[`pkg/storage/pebble_iterator.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJg8cg9w4GYayEv-r2-52:-NRe8a-n2WFTsAceb2UW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d725ca9a6e8b43dae337103a2406e48ff9ea347b/pkg/storage/pebble_iterator.go#L52)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

and it can't be nil, right?
</blockquote></details>

Done.

**Comment by nicktrav** [comment]
Thoughts on adding some / all of these metrics to the telemetry event pipeline? Would they be helpful for observing long term trends or identifying areas for investigation / improvement?

**Review by tbg — APPROVED** [review]
> eg, running this benchmark on my gceworker:

I dug a little deeper on this benchmark. I don't think a parallelism of 64 is sufficient - since this is in the read hot path, we have to think about 1000s of ops/sec, so I extended to high concurrencies (and thinned it out a bit).
Second, I think the benchmark was doing it wrong? It was running `parallelism*b.N` operations but the benchmark harness assumed it was `b.N`. So we'd really have to divide all reported ops/sec numbers by `parallelism`. I cleaned that all up a bit (using `b.RunParallel` in the benchmark and running only `b.N` ops each). I also added a sharded mutex (GOMAXPROCS shards). The resulting numbers  (gceworker) are... surprising, the standard mutex actually beats the other options and actually gets ... better? It's almost as though queuing up behind a single mutex is more efficient than shorter queues across multiple mutexes. I wonder if there's something smart going on in the scheduler, or something dumb in my code.



https://github.com/tbg/goplay/blob/master/mutex/mutex_test.go
https://docs.google.com/spreadsheets/d/1B9g1ZfzjhpZefJr9RlZWKJ3XvNiokHW10M8q-Eb7--A/edit?usp=sharing
 
I suppose if this new mutex does cause an issue it is best observed in roachperf rather than embarking on discussion too focused on microbenchmarks.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-:-NRhKvzM9FdgTng8YsRC:b-781vtx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick and @jbowens)

___
*[`pkg/kv/kvserver/metrics.go` line 2473 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJhCgwgFH1T-Hza-r2-2473:-NRhMuOC7-bbFWRtDWDV:b-zetcuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d725ca9a6e8b43dae337103a2406e48ff9ea347b/pkg/kv/kvserver/metrics.go#L2473)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I was expecting them all to be counters. They are all gauges. What's going on?
</blockquote></details>

Could you add a comment that these are cumulative values, but tracked as gauges because the cumulative counters mostly live in pebble (though this isn't always true, it's more convenient to stick to the pattern throughout), at the top of this block (i.e. not on each metric)?

I also wonder (not for this PR) if we can't use counters here regardless. A gauge and a counter get different prometheus metric types, and there are consequences of this type. For example, plotting a counter on grafana, it will remind you that this is a counter and propose looking at the `rate()`. Having a gauge that is really a counter is unintuitive.

I think the only reason we're using gauges here is because we don't have a `.Update` method on `Counter`. I think this could be added and all of these metrics ported. Filed https://github.com/cockroachdb/cockroach/issues/99922, perhaps mention it in the comment requested above.

File: pkg/storage/pebble.go [hunk 6]
```diff
@@ -1338,13 +1357,13 @@ func (p *Pebble) NewMVCCIterator(iterKind MVCCIterKind, opts IterOptions) MVCCIt
 		return maybeWrapInUnsafeIter(iter)
 	}
 
-	iter := newPebbleIterator(p.db, opts, StandardDurability)
+	iter := newPebbleIterator(p.db, opts, StandardDurability, p.aggregateIterStats)
```
Comment (nvanbenschoten) [inline pkg/storage/pebble.go:1360]: Drive-by comment: does passing this method reference to `newPebbleIterator` result in a new heap allocation? If so, consider replacing `func(IteratorStats)` with an interface that is implemented by `Pebble`. `p *Pebble` is already on the heap, so that approach would not heap allocation.

**Review by RaduBerinde — COMMENTED** [review]
Not that SetParallelism sets the number of goroutines used by RunParallel to p*GOMAXPROCS so the parallelism you are looking at is way too high

I have a feeling that the atomic would fare relatively worse if the values were not packed close together in 1-2 cachelines.

It might be more intuitive to show a graph of throughput (total number of ops per second). That one should grow ~linearly up to the number of CPUs then level off and start going down a bit. I think it would also show that the number is so much higher than what we're actually doing in CRDB that this is mostly academic :joy: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-:-NRhhbidCWWc-VnELGhr:b435umo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @jbowens, and @tbg)

**Review by jbowens — COMMENTED** [review]
> It was running parallelism*b.N operations but the benchmark harness assumed it was b.N. So we'd really have to divide all reported ops/sec numbers by parallelism

Makes sense, but it doesn't impact the relative numbers between mutex and atomics.

>  I don't think a parallelism of 64 is sufficient - since this is in the read hot path, we have to think about 1000s of ops/sec,

There would be need to millions of iterators closing in parallel to see 1000s of iterators aggregating stats in parallel. Since the critical section is tiny, my expectation is that higher parallelism beyond the expected number of CPU threads won't matter. The probability of any individual `aggregateIterStats` call being context switched away while in the critical section is incredibly small, and IIRC the golang mutex has an optimistic spin-loop fast path for very short critical sections like this.

I think we see the effect of the cache line invalidation on every individual atomic increment. At the extreme parallelism counts, every increment can expect a cache miss. In contrast, the mutex suffers the cost of the write barrier and cache miss once, on mutex acquisition.

FWIW, iterator construction also acquires an Engine-wide RWMutex in order to acquire the current version of the LSM. We've never seen contention there. (We typically see it on the block cache shards, which see much higher ops/sec and are sharded 2*num procs: cockroachdb/pebble#1997).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-:-NRhoN5lAZwi_j5yZnyC:b8s4978)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @nvanbenschoten, and @tbg)

___
*[`pkg/kv/kvserver/metrics.go` line 2473 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJhCgwgFH1T-Hza-r2-2473:-NRhsqM01Yb9r2ynG4k5:b8x0d9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d725ca9a6e8b43dae337103a2406e48ff9ea347b/pkg/kv/kvserver/metrics.go#L2473)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could you add a comment that these are cumulative values, but tracked as gauges because the cumulative counters mostly live in pebble (though this isn't always true, it's more convenient to stick to the pattern throughout), at the top of this block (i.e. not on each metric)?

I also wonder (not for this PR) if we can't use counters here regardless. A gauge and a counter get different prometheus metric types, and there are consequences of this type. For example, plotting a counter on grafana, it will remind you that this is a counter and propose looking at the `rate()`. Having a gauge that is really a counter is unintuitive.

I think the only reason we're using gauges here is because we don't have a `.Update` method on `Counter`. I think this could be added and all of these metrics ported. Filed https://github.com/cockroachdb/cockroach/issues/99922, perhaps mention it in the comment requested above.
</blockquote></details>

Added a comment here, and updated the existing comment above the struct comments to reference #99922.
___
*[`pkg/storage/pebble.go` line 1360 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-NRZkyJhCgwgFH1T-Hzb-r4-1360:-NRhtatVEFtt-j9nvvmj:b87e77t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d901bc0efad4f554a086a815dbe5684dec7fbb41/pkg/storage/pebble.go#L1360)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Drive-by comment: does passing this method reference to `newPebbleIterator` result in a new heap allocation? If so, consider replacing `func(IteratorStats)` with an interface that is implemented by `Pebble`. `p *Pebble` is already on the heap, so that approach would not heap allocation.
</blockquote></details>

Good catch, it does. Fixed.

**Comment by tbg** [comment]
Thanks for the comments, I agree that these numbers probably don't mean much. But, since I was at it anyway, I cleaned things up and look, nothing seems to matter any more. The previous discrepancies were caused by the bench runner's attempt to measure the needed number of iterations. Since spinning up `p` goroutines is a fixed cost, at some point this went off the rails and used only a single run, which now benchmarked the `go` keyword more than anything else. With this straightened out (`-benchtime 100000x`):

image

**Comment by tbg** [comment]
^-- hiding the above, @RaduBerinde pointed out the numbers are still garbage - it's hard to get something useful out of the go bench harness, @jbowens' initial attempt was still the best shot. I went looking upstream too but they only have a very basic mutex benchmark. I suppose it doesn't matter, since we're confident there isn't going to be noticeable contention. Sorry about the noise!

**Review by tbg — APPROVED** [review]
LGTM (but note the eng-wide ask to not bors anything right now)

> Reviewed 10 of 10 files at r5, 11 of 11 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-:-NRikvRZ3DO5QcfzITUZ:b-os85cj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick and @nvanbenschoten)

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 5 files at r1, 8 of 10 files at r5, 11 of 11 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99726#-:-NRnGag_BArjPiMmtCKj:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick and @nvanbenschoten)

**Comment by jbowens** [comment]
TFTRs!

bors r+

**Comment by jbowens** [comment]
bors retry

**Comment by jbowens** [comment]
bors retry
